### PR TITLE
Auth feature

### PR DESCRIPTION
**Pull request by jfranciscosousa**
Placeholder PR for the meantime

### CODE REVIEW

File: src/backend/test/support/factories.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const faker = require("faker");
+const { Factory } = require("rosie");
+const { sample } = require("lodash");
+const Firefighter = require("../../models/firefighter");
+
+const firefightersFactory = new Factory().attrs({
+  name: faker.name.firstName,
+  email: faker.internet.email(),
+  password: "foobar",
+  status: sample("active", "inactive", "busy")
```
Comment (luistorres) [inline src/backend/test/support/factories.js:10]: We are missing one attr that is `dutyType`

Comment (jfranciscosousa) [reply src/backend/test/support/factories.js:10]: Not need for the tests though, wanna add it?

Comment (jfranciscosousa) [reply src/backend/test/support/factories.js:10]: Besides, dont write reviews here, this is a long standing PR :stuck_out_tongue:

Comment (luistorres) [reply src/backend/test/support/factories.js:10]: I know, but had nowhere else to comment. And duty is not working in the app :/

**Comment by jfranciscosousa** [comment]
@luistorres anything else you need?

**Comment by luistorres** [comment]
Maybe assuring that the `dutyType` is working as I don't see it anywhere besides the seeds.
Besides that, do we have endpoints for changing password and login out? If so, all is good I think

**Comment by jfranciscosousa** [comment]
We don't have a logout route, we are using stateless tokens so the only possibility would be to implement a token blacklist. As I told you before the best possible way would be to have short-lived tokens (30 mins or so) and to refresh the token on the client-side every 30 mins.

**Comment by jfranciscosousa** [comment]
About the `dutyType` I really can't understand the problem, I didn't meddle with it at all.

**Comment by luistorres** [comment]
30 mins is really low. I would say at least 7 days or so, and every visit to the app in FE could try to refresh it (for instance). Token expiration in this app is just a pain, so the more we can delay it, the better.
Logout is needed just in case you want to change user...we shouldn't just clean the token in FE, that is not the best practice,  though for now it is enough.

There is no problem with `dutyType`, I'm just saying we can have tests, as we have for all the other things.

**Comment by jfranciscosousa** [comment]
To the first question: I'll add session-based auth then.

To the other question: I'll add the duty type then.

**Comment by jfranciscosousa** [comment]
Forget the session-based auth on node.js, I can't do it (no time for it).

About cleaning the token... JWT is a stateless authentication method, there is no "logout" and it's used for simple APIs. It's easier to test and develop with on Node, rather than using cookies and sessions and whatnot. And even with session and cookie-based mechanisms, you have the same problems that you have with stateless, in that you have no way of securing your app if a hacker manages to capture a token on requests.

If you use HTTPS 99.99% no problems with people getting the token (unless they hack HTTPS on itself). The only sure way people can hack something is if they open the user's browser and fetch the token from localStorage (they would need to actually steal a computer or a phone). Native apps use SecureStorage APIs on Android and iOS, but then again physical access means you are f****.

On a breach, you can change the `JWT_SIGN_KEY` variable and all of the tokens are invalidated.

As of now and the future. 7 days of the expiry time. Every time the user opens the app you refresh the token. Check out the spec of OAuth. They work by having a short-lived token and a long-lived refresh token (the second one lives on the browser). You use the short-lived token in exchanges with a server and then you use the refresh token to ask for a new token every now and then. 

Currently, I don't plan on implementing refresh tokens, I would just create a route that returns a newly built JWT. These things are hard to make and test on part-time. Perhaps moving to Rails would be a good idea but now it's too late.

If this would be a native app, you could just store the token in SecureSTorage and use biometrics to fetch it.

**Comment by jfranciscosousa** [comment]
Things to do:
* Add refresh token route
* Add dutyType to tests
* Add role-based authorization (admin vs regular user)
* Add authorization to the `socket.io` module (should be admin only)

**Comment by luistorres** [comment]
I'm good with everything. If we are able to change `JWT_SIGN_KEY ` and invalidate everything, can we have the expiration be something like 3 months then? Even 7 days seem too little when guys are expected to open this like once a month or something.

**Comment by luistorres** [comment]
We need 3 degrees of authorization actually:
- user only gets to new availability screen, so basically can only change it's own status.
- dashboard (only sees dashboard) - can manage everyone status and dutyType
- admin - sees dashboard (change status and dutyType), can add/remove users, can send notifications (to be done yet)

**Comment by jfranciscosousa** [comment]
@luistorres given new developments, can I close this and also archive the repo?

**Pull request closed without merging**